src/main/java/org/wildfly/security/password/spec/PasswordSpecEncoder.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * JBoss, Home of Professional Open Source
+ *
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.password.spec;
+
+import org.wildfly.security.util.ByteIterator;
+
+import java.nio.ByteBuffer;
+import java.security.spec.InvalidKeySpecException;
+
+/**
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class PasswordSpecEncoder {
+
+    public String encode(PasswordSpec passwordSpec) throws InvalidKeySpecException {
+        StringBuilder encoded = new StringBuilder();
+
+        if (SaltedHashPasswordSpec.class.isInstance(passwordSpec)) {
+            SaltedHashPasswordSpec spec = (SaltedHashPasswordSpec) passwordSpec;
```

Comment:
If you do not write a lead byte that identifies the spec type, you will not be able to decode it again.